linux/mcompile/keymap/mcompile.cpp

```diff
@@ -22,77 +22,68 @@
-//
-// m-to-p.cpp : Defines the entry point for the console application.
-//
-// Note: this program deliberately leaks memory as it has a very short life cycle and managing the memory allocations
-// for the subcomponents of the compiled keyboard is an unnecessary optimisation. Just so you know.
-//
 
-#include "pch.h"
-#include <vector>
+#include "mcompile.h"
 
-#include <stdio.h>
-#include <stdarg.h>
-#include <varargs.h>
+//------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------
 
-BOOL DoConvert(LPKEYBOARD kbd, PWSTR kbid, BOOL bDeadkeyConversion);
-BOOL SaveKeyboard(LPKEYBOARD kbd, PWSTR filename);
-bool ImportRules(WCHAR *kbid, LPKEYBOARD kp, std::vector<DeadkeyMapping> *FDeadkeys, BOOL bDeadkeyConversion);   // I4353   // I4327
-BOOL ConvertKeyboardToUnicode(LPKEYBOARD kbd);   // I4273
-int run(int argc, wchar_t * argv[]);
 
-std::vector<DeadkeyMapping> FDeadkeys;   // I4353
+KMX_BOOL KMX_DoConvert(LPKMX_KEYBOARD kbd, KMX_BOOL bDeadkeyConversion, gint argc, gchar *argv[]);
 
-#define KEYMAN_SENTRY_LOGGER_DESKTOP_ENGINE_MCOMPILE KEYMAN_SENTRY_LOGGER_DESKTOP_ENGINE ".mcompile"
+bool KMX_ImportRules( LPKMX_KEYBOARD kp,v_dw_3D &All_Vector, GdkKeymap **keymap,std::vector<KMX_DeadkeyMapping> *KMX_FDeadkeys, KMX_BOOL bDeadkeyConversion); // I4353   // I4327
 
-int wmain(int argc, wchar_t * argv[])
-{
-  return keyman_sentry_wmain(false, KEYMAN_SENTRY_LOGGER_DESKTOP_ENGINE_MCOMPILE, argc, argv, run);
-}
+std::vector<KMX_DeadkeyMapping> KMX_FDeadkeys; // I4353
 
-int run(int argc, wchar_t * argv[])
-{
-  if(argc < 3 || (argc < 5 && wcscmp(argv[1], L"-u") != 0)) {   // I4273
-    printf(
-         "Usage: mcompile -u infile.kmx outfile.kmx\n"
-         "       mcompile [-d] infile.kmx kbdfile.dll kbid outfile.kmx\n"
-         "  With -u parameter, converts keyboard from ANSI to Unicode\n"
-         "  Otherwise, mcompile converts a Keyman mnemonic layout to a\n"
-         "  positional one based on the Windows keyboard\n"
-         "  layout file given by kbdfile.dll\n\n"
-         "  kbid should be a hexadecimal number e.g. 409 for US English\n"
-         "  -d   convert deadkeys to plain keys\n");   // I4552
+// Note: max is not a standard c api function or macro
+#ifndef max
+#define max(a,b)            (((a) > (b)) ? (a) : (b))
+#endif
 
-    return 1;
-  }
+#define _countof(a) (sizeof(a)/sizeof(*(a)))
 
-  if(wcscmp(argv[1], L"-u") == 0) {   // I4273
-    wchar_t *infile = argv[2], *outfile = argv[3];
+#if defined(_WIN32) || defined(_WIN64)
+  int wmain(int argc, wchar_t* argv[]) {
+    std::vector<std::u16string> str_argv_16 = convert_argvW_to_Vector_u16str( argc, argv);
+    run(argc, str_argv_16);
 
-    LPKEYBOARD kmxfile;
+#else  // LINUX
+  int main(int argc, char* argv[]) {
+    std::vector<std::u16string> str_argv_16 = convert_argv_to_Vector_u16str(argc, argv);
+    run(argc, str_argv_16, argv);
```

Comment:
On Linux there is no need to pass UTF-16 strings for filenames. Linux uses UTF-8 and so we can directly use `char *` without converting back and forth from/to UTF-16. This will simplify the code quite a bit.

Reply:
OK-done #11887